infra/Pos/DHT/Real/Real.hs

```diff
@@ -2,57 +2,62 @@
 
 import           Control.Concurrent.STM    (newTVar, readTVar, writeTVar)
 import           Data.Binary               (decode)
+import qualified Data.ByteString.Char8     as B8 (unpack)
 import qualified Data.ByteString.Lazy      as BS
 import qualified Data.HashMap.Strict       as HM
 import           Data.List                 (intersect, (\\))
 import           Formatting                (build, int, sformat, shown, (%))
 import           Mockable                  (Async, Catch, Mockable, MonadMockable,
                                             Promise, Throw, bracket, catchAll, fork,
-                                            killThread, throw, waitAnyUnexceptional)
+                                            killThread, throw, waitAnyUnexceptional,
+                                            catch)
 import qualified Network.Kademlia          as K
 import           Serokell.Util             (ms, sec)
 import           System.Directory          (doesFileExist)
 import           System.Wlog               (WithLogger, logDebug, logError, logInfo,
-                                            logWarning, usingLoggerName)
+                                            logWarning, usingLoggerName,
+                                            HasLoggerName (modifyLoggerName))
 
 import           Pos.Binary.Class          (Bi (..))
 import           Pos.Binary.Infra.DHTModel ()
 import           Pos.DHT.Constants         (enhancedMessageBroadcast,
                                             neighborsSendThreshold)
-import           Pos.DHT.Model.Class       (DHTException (..), MonadDHT (..),
-                                            withDhtLogger)
-import           Pos.DHT.Model.Types       (DHTData, DHTKey, DHTNode (..), randomDHTKey)
-import           Pos.DHT.Model.Util        (joinNetworkNoThrow)
-import           Pos.DHT.Real.Types        (DHTHandle, KademliaDHT (..),
-                                            KademliaDHTInstance (..),
+import           Pos.DHT.Model.Types       (DHTData, DHTKey, DHTNode (..), randomDHTKey,
+                                            DHTException (..))
+import           Pos.DHT.Real.Types        (KademliaDHTInstance (..),
                                             KademliaDHTInstanceConfig (..))
 import           Pos.Util.TimeLimit        (runWithRandomIntervals')
 import           Pos.Util.TimeWarp         (NetworkAddress)
 
 kademliaConfig :: K.KademliaConfig
 kademliaConfig = K.defaultConfig { K.k = 16 }
 
--- | Run KademliaDHT
-runKademliaDHT
+foreverRejoinNetwork
     :: ( MonadMockable m
        , Eq (Promise m (Maybe ()))
        , MonadIO m
        , WithLogger m
        )
-    => KademliaDHTInstance -> KademliaDHT m a -> m a
-runKademliaDHT inst action = runReaderT (unKademliaDHT action') inst
+    => KademliaDHTInstance
+    -> m a
+    -> m a
+foreverRejoinNetwork inst action = action'
   where
+    -- TODO this is not correct. `killThread` is interruptible, so the
```

Comment:
Not sure I'm correct, but you've already solved this kind of issue using `withAsync`.